### PR TITLE
[#Wave-Control: Add NetSuite]: Settings Configuration in NewDot: Import Mapping

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

Covers the implementation of [Import mappings page](https://docs.google.com/document/d/1WubNv_VAv78IxG4FKsi9aS0pWESfWvUYbqBAAy5b4bc/edit#heading=h.9bcymmz9e7bu) - Departments, Classes, Locations, Customers, Projects.

This is one of the FE PRs for NetSuite without the proper auth flow integrated. Hence, to get this working we'll need to have a few pre-requisites:

1. Collect workspace with accounting enabled.
2. `NetSuite` beta enabled for the user.  You can return true in the [canUseNetSuiteIntegration](https://github.com/Expensify/App/blob/6294e7d3462abe4124e6486de6702727b78a2c2c/src/libs/Permissions.ts#L51-L53) method.
3. Setup the Netsuite data. You can either connect the account using OldDot or populate the data linked in the comment [here](https://github.com/Expensify/App/issues/43435#issuecomment-2159929615) for the given workspace as `connections: {netsuite: <linkedjson>}`. 
> One way I did the setup is calling `Onyx.merge` with `useEffect` on any FE component.
> ```
>    useEffect(() => {
>        Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY}76A10AB364F95DA0`, {"connections":{"netsuite":{"accountID":"TSTDRV1668486", ....})
>    }, []);
>
> ``` 

### Fixed Issues

$ https://github.com/Expensify/App/issues/43437
PROPOSAL:

### Tests

1. Open the app with the pre-setup workspace.
2. Go to `Accounting` from the LHS
3. Verify that you're able to see `Net Suite` in the active connections along with the `Oracle NetSuite` logo.
4. Click on the `Import` button.
5. Verify that the list of options appear as per the linked design doc.
6. Attempt the following for each of the import fields - Departments, Classes, Locations, Customers, Projects
      a. Clicking the menu item should open a new page on the RHP.
      b. It should contain title corresponding to the field. ex. Departments.
      c. It should have the subtitle based on the design doc.
      d. For departments, classes, and locations it should show 3 options - `NetSuite employee default`, `Tags` and `Report field` along with the description and a footer based on the selected option. For `Customers` and `Projects`, we should only show `Tags` and `Report field`.
      e. Clicking on each of the options should save the config for the field, and show up in the `Import` page.
7. For `Projects` we have additional toggle on the screen - `Cross-subsidiary customers`. Clicking should toggle to `config?.syncOptions?.crossSubsidiaryCustomers` in the policy data.
8. In case of the API error, there should be an error message below the switch or each of the menu item.     

- [ ] Verify that no errors appear in the JS console

### Offline tests

Same as Test steps, with the greyed out menu item in case of optimistic pending updates.

### QA Steps

Same as Test Steps.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/3069065/4c477cf6-429a-4a9d-93fc-f34b47783025

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/3069065/fd4d3a61-cd2c-4b07-bce8-9e90fb812a63

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/3069065/ef6fa630-d760-4202-99a3-536e0f8adb8f

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/3069065/c55ec1db-56ee-420a-b3dc-d0a5d492d071

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

**Happy Flow**

https://github.com/Expensify/App/assets/3069065/811d0afa-5ea4-4fc1-832e-7448e76c6645

**Spanish Translations and Errors**
https://github.com/Expensify/App/assets/3069065/f3ce1aa4-352f-4a37-acec-3dc16824562b

**New Screen for Customers / projects**

https://github.com/Expensify/App/assets/3069065/909350ae-e94d-45aa-9dae-c61e26bf191f

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/3069065/4345b3d8-9981-4ee7-adb2-f8d96b5738c6

https://github.com/Expensify/App/assets/3069065/8c58eba0-2f0f-4578-a939-f9a7ad8d8d6a

</details>

### CODE REVIEW

**Comment by mananjadhav** [comment]
@hungvu193 Please ignore this PR.

@shubham1206agra @yuwenmemon @Expensify/design This is ready for review.

**Comment by shawnborton** [comment]
Generally looking pretty good! 

On this screen, I would think the text above the list options would not be gray and would have slightly more padding below and maybe less padding above?
CleanShot 2024-07-03 at 08 08 53@2x

**Comment by mananjadhav** [comment]
@shubham1206agra @yuwenmemon This is now ready for review.

File: src/CONST.ts [hunk 1]
```diff
@@ -1373,10 +1373,15 @@ const CONST = {
         PROVINCIAL_TAX_POSTING_ACCOUNT: 'provincialTaxPostingAccount',
         ALLOW_FOREIGN_CURRENCY: 'allowForeignCurrency',
         EXPORT_TO_NEXT_OPEN_PERIOD: 'exportToNextOpenPeriod',
-        IMPORT_FIELDS: ['departments', 'classes', 'locations', 'customers', 'jobs'],
+        IMPORT_FIELDS: ['departments', 'classes', 'locations'],
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/CONST.ts:1376]: Didn't see any advantage for that. I am anyway looping through these and using as type.

**Comment by mananjadhav** [comment]
> Screenshot 2024-07-04 at 6 36 04 PM
> Text under `Customer / projects` is different from current mock.


@shubham1206agra You mean we're showing `Imported` whereas the mock up shows `displayed` or something else? @yuwenmemon I am using the existing translations, can you confirm if we need to show `display`? I did see one more issue that `Imported` should be small case.

File: src/pages/workspace/accounting/netsuite/import/NetSuiteImportCustomersOrProjectsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    const updateMapping = useCallback(
+        (importField: ImportField, isEnabled: boolean) => {
+            let newValue;
+            if (!isEnabled) {
+                // if the import is off, then we send default as the value for mapping
+                newValue = CONST.INTEGRATION_ENTITY_MAP_TYPES.NETSUITE_DEFAULT;
+            } else {
+                // when we enable any field, and if the other one already has a value set, we should set that,
+                const otherFieldValue = importField === 'jobs' ? config?.syncOptions?.mapping?.customers : config?.syncOptions?.mapping?.jobs;
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/pages/workspace/accounting/netsuite/import/NetSuiteImportCustomersOrProjectsPage.tsx]: yeah missed this. Thanks for highlighting.

**Comment by mananjadhav** [comment]
@shubham1206agra I've fixed the content overflowing as well that you reported.

web-import-truncate-issue

**Review by shubham1206agra — APPROVED** [review]

File: src/libs/PolicyUtils.ts [hunk 3]
```diff
@@ -536,6 +537,32 @@ function canUseProvincialTaxNetSuite(subsidiaryCountry?: string) {
...
+    const importMapping = policy?.connections?.netsuite?.options?.config?.syncOptions?.mapping;
+    if (!importMapping?.customers && !importMapping?.jobs) {
+        return undefined;
+    }
+    const importFields: string[] = [];
+    const importCustomer = importMapping?.customers ?? CONST.INTEGRATION_ENTITY_MAP_TYPES.NETSUITE_DEFAULT;
+    const importJobs = importMapping?.jobs ?? CONST.INTEGRATION_ENTITY_MAP_TYPES.NETSUITE_DEFAULT;
+    const importedValue = importMapping?.customers !== CONST.INTEGRATION_ENTITY_MAP_TYPES.NETSUITE_DEFAULT ? importCustomer : importJobs;
+
+    if (!importedValue || importedValue === CONST.INTEGRATION_ENTITY_MAP_TYPES.NETSUITE_DEFAULT) {
```
Comment (yuwenmemon) [inline src/libs/PolicyUtils.ts:550]: Based on the logic above `importedValue` will always be defined (or truthy).

Comment (yuwenmemon) [reply src/libs/PolicyUtils.ts:550]: Simple would just be to say: if both are `NETSUITE_DEFAULT`, return `undefined` early.

File: src/libs/actions/connections/NetSuiteCommands.ts [hunk 4]
```diff
@@ -283,6 +283,125 @@ function updateNetSuiteSubsidiary(policyID: string, newSubsidiary: SubsidiaryPar
...
+            commandName = WRITE_COMMANDS.UPDATE_NETSUITE_CLASSES_MAPPING;
+            break;
+        case 'locations':
+            commandName = WRITE_COMMANDS.UPDATE_NETSUITE_LOCATIONS_MAPPING;
+            break;
+        case 'customers':
+            commandName = WRITE_COMMANDS.UPDATE_NETSUITE_CUSTOMERS_MAPPING;
+            break;
+        case 'jobs':
+            commandName = WRITE_COMMANDS.UPDATE_NETSUITE_JOBS_MAPPING;
```
Comment (yuwenmemon) [inline src/libs/actions/connections/NetSuiteCommands.ts:396]: Just a note is that we'll change these later down the line once we have a new API for this.

**Review by yuwenmemon — CHANGES_REQUESTED** [review]

File: src/CONST.ts [hunk 5]
```diff
@@ -1396,6 +1396,11 @@ const CONST = {
             EXPORT_VENDOR_BILLS_TO: 'exportVendorBillsTo',
             EXPORT_JOURNALS_TO: 'exportJournalsTo',
             SYNC_TAX: 'syncTax',
+            CROSS_SUBSIDIARY_CUSTOMERS: 'crossSubsidiaryCustomers',
+            MAPPING: {
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/CONST.ts:1400]: Well I used the the exact key that we have in the config. I can update.

File: src/libs/PolicyUtils.ts (hunk 3 above)
Comment (mananjadhav) [reply src/libs/PolicyUtils.ts]: Updated.

**Review by yuwenmemon — APPROVED** [review]

**Pull request merged**